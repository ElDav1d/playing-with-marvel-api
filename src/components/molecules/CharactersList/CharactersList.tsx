export interface Thumbnail {
  path: string;
  extension: string;
}

export interface Character {
  description: string;
  id: number;
  name: string;
  modified: string;
  thumbnail: Thumbnail;
}

export interface CharactersListProps {
  characters: Character[];
}

const CharactersList = ({ characters }: CharactersListProps) => {
  return (
    <ul>
      {characters.map((character) => (
        <li key={character.id}>
          <h2>{character.name}</h2>
          <p>
            <small>
              <strong>thumbnail: </strong>
              {character.thumbnail.path}
            </small>
          </p>
          <p>
            <small>
              <strong>modified: </strong>
              {character.modified}
            </small>
          </p>
          {character.description ? (
            <p>
              <strong>DESCRIPTION: </strong>
              {character.description}
            </p>
          ) : (
            <p>
              <strong>DESCRIPTION NOT AVAILABLE</strong>
            </p>
          )}
        </li>
      ))}
    </ul>
  );
};

export default CharactersList;
