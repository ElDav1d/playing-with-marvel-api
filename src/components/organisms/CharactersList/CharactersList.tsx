import { CharacterItem } from '@/components/pages/Characters/interfaces/characters';
import CharacterListItem from '@/components/molecules/CharacterListItem/CharacterListItem';
export interface CharactersListProps {
  characters: CharacterItem[];
}

const CharactersList = ({ characters }: CharactersListProps) => {
  return (
    <ul className='grid gap-3 grid-flow-row grid-cols-auto-min-max-120-auto md:grid-cols-auto-min-max-195-auto'>
      {characters.map(({ id, name, thumbnail, description }) => (
        <CharacterListItem
          key={id}
          id={id}
          name={name}
          thumbnail={thumbnail}
          description={description}
        />
      ))}
    </ul>
  );
};

export default CharactersList;
