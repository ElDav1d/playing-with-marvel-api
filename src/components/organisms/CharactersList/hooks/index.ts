export { default as useFilteredCharacters } from './useFilteredCharacters';
