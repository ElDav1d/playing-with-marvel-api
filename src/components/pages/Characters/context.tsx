import { useContext, createContext, Ref } from 'react';

interface Context {
  refProp: Ref<HTMLLIElement> | null;
  counter: number | null;
  setCounter: React.Dispatch<React.SetStateAction<number>> | null;
}

const CharactersContext = createContext<Context>({
  refProp: null,
  counter: null,
  setCounter: null,
});

const useCharactersContext = () => {
  const context = useContext(CharactersContext);

  if (context === undefined) {
    throw new Error('useCharactersContext was used outside of its Provider!');
  }

  return context;
};

export { useCharactersContext };
export default CharactersContext;
