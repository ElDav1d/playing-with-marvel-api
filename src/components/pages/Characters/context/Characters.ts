import { createContext } from 'react';
import combineReducers from 'react-combine-reducers';
import filtersReducer, {
  FiltersState,
  FiltersAction,
  initialFiltersState,
} from '../reducers/FiltersReducer';
import orderReducer, {
  IOrderState,
  IOrderAction,
  initialOrderState,
} from '../reducers/OrderReducer';
import searchStringReducer, {
  SearchStringAction,
  SearchStringState,
  initialSearchStringState,
} from '../reducers/SearchStringReducer';

type Action = SearchStringAction | FiltersAction | IOrderAction;

interface IState {
  searchString: SearchStringState;
  filters: FiltersState;
  order: IOrderState;
}

type CombinedReducer = (state: IState, action: Action) => IState;

type Dispatch = (action: Action) => void;

export const [combinedReducer, initialCombinedState] = combineReducers<CombinedReducer>({
  searchString: [searchStringReducer, initialSearchStringState],
  filters: [filtersReducer, initialFiltersState],
  order: [orderReducer, initialOrderState],
});

const CharactersContext = createContext<
  { charactersContextState: IState; charactersContextDispatch: Dispatch } | undefined
>(undefined);

export default CharactersContext;
