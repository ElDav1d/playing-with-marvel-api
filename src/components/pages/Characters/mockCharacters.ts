import { CharacterItem } from '@/interfaces/globals';

export const mockCharacters: CharacterItem[] = [
  {
    id: 1011176,
    name: 'Animal',
    description:
      'Xorn, whose mutation manifested as a tiny, powerful star inside his head, was brainwashed into believing he was the reincarnation of Magneto.',
    modified: '1969-12-31T19:00:00-0500',
    thumbnail: {
      path: 'http://i.annihil.us/u/prod/marvel/i/mg/2/80/4c002f35c5215',
      extension: 'jpg',
    },
  },
  {
    id: 1011177,
    name: 'Braineater',
    description:
      'Xorn, whose mutation manifested as a tiny, powerful star inside his head, was brainwashed into believing he was the reincarnation of Magneto.',
    modified: '1969-12-31T19:00:00-0500',
    thumbnail: {
      path: 'http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available',
      extension: 'jpg',
    },
  },
  {
    id: 1011178,
    name: 'Crusher',
    description: '',
    modified: '1969-12-31T19:00:00-0500',
    thumbnail: {
      path: 'http://i.annihil.us/u/prod/marvel/i/mg/2/80/4c002f35c5215',
      extension: 'jpg',
    },
  },
  {
    id: 1011179,
    name: 'Destructor',
    description: '',
    modified: '1969-12-31T19:00:00-0500',
    thumbnail: {
      path: 'http://i.annihil.us/u/prod/marvel/i/mg/b/40/image_not_available',
      extension: 'jpg',
    },
  },
];
