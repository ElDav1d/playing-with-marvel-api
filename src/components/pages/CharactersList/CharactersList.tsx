export interface Thumbnail {
  path: string
  extension: string
}

export interface Character {
  id: number
  name: string
  description: string
  thumbnail: Thumbnail
}

export interface CharactersListProps {
  characters: Character[]
}

const CharactersList = ({ characters }: CharactersListProps) => {
  return (
    <ul>
      {characters.map((character) => (
        <li key={character.id}>{character.name}</li>
      ))}
    </ul>
  )
}

export default CharactersList;