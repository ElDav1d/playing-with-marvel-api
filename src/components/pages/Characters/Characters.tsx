import { ChangeEvent, useEffect, useRef, useState, useMemo } from 'react';
import { useInView } from 'react-intersection-observer';
import { useCharacters, useDebounce } from './hooks';
import { CharacterItem, FetchingOrder, FilterCriteria } from './interfaces/characters';
import CharactersList from '@/components/organisms/CharactersList/CharactersList';
import {
  SELECT_ORDER_LITERALS,
  EMPTY_DATA_LITERAL_LIST,
  LOADER_SIZE,
  MARVEL_RED,
  MAX_FETCH_CHARACTERS_DEFAULT,
  MAX_FETCH_CHARACTERS_OPTIM,
  MAX_FETCH_CHARACTERS_TOP,
  REGEX_IMAGE_PATH,
} from '@/utils/constants';
import Header from '@/components/organisms/Header';
import Footer from '@/components/organisms/Footer';
import Container from '@/components/organisms/Container';
import { RingLoader } from 'react-spinners';
import SideDrawer from '@/components/organisms/SideDrawer';
import CharactersControlPanel from '@/components/organisms/CharactersControlPanel';

const Characters = () => {
  const [searchInput, setSearchInput] = useState<string>('');
  const [searchString, setSearchString] = useState<string>('');
  const [order, setOrder] = useState<FetchingOrder>(FetchingOrder.NAME_AZ);
  const [onClearData, setOnClearData] = useState(false);
  const [filters, setFilters] = useState<FilterCriteria[]>([]);
  const [onClearFilters, setOnClearFilters] = useState(false);
  const maxCharactersRef = useRef(MAX_FETCH_CHARACTERS_DEFAULT);

  const { isError, characters, fetchNextPage, hasNextPage, isFetching, isFetchingNextPage } =
    useCharacters({ maxCharacters: maxCharactersRef.current, searchString, order, onClearData });

  const { ref, inView } = useInView({
    threshold: 0.1,
  });

  useEffect(() => {
    if (inView) {
      fetchNextPage();
    }
  }, [inView]);

  useEffect(() => {
    if (onClearData) {
      setOnClearData(false);
    }
  }, [order, searchString]);

  useDebounce(searchInput, 500, () => setSearchString(searchInput));

  const filterLiterals = ['With Image', 'With Description'];

  const filteredCharacters = useMemo(() => {
    const hasImage = (path: string) => !REGEX_IMAGE_PATH.test(path);
    const hasDescription = (description: string) => description && description !== ' ';

    let maxCharacters;
    let filterCallback;

    switch (true) {
      case filters.includes(FilterCriteria.IMAGE) && filters.includes(FilterCriteria.DESCRIPTION):
        maxCharacters = MAX_FETCH_CHARACTERS_TOP;
        filterCallback = (character: CharacterItem) =>
          hasImage(character.thumbnail.path) && hasDescription(character.description);
        break;

      case filters.includes(FilterCriteria.IMAGE):
        maxCharacters = MAX_FETCH_CHARACTERS_OPTIM;
        filterCallback = (character: CharacterItem) => hasImage(character.thumbnail.path);
        break;

      case filters.includes(FilterCriteria.DESCRIPTION):
        maxCharacters = MAX_FETCH_CHARACTERS_OPTIM;
        filterCallback = (character: CharacterItem) => hasDescription(character.description);
        break;

      default:
        maxCharactersRef.current = MAX_FETCH_CHARACTERS_DEFAULT;
        return characters;
    }

    maxCharactersRef.current = maxCharacters;
    return characters.filter(filterCallback);
  }, [characters, filters]);

  const orderHandler = (event: ChangeEvent<HTMLSelectElement>): void => {
    const value = event.target.value as FetchingOrder;

    setOnClearData(true);
    setOrder(value);
  };

  const hasListDefaults =
    searchInput === '' && order === FetchingOrder.NAME_AZ && filters.length === 0;

  const handleClear = () => {
    setSearchInput('');
    setSearchString('');
    setOrder(FetchingOrder.NAME_AZ);
    setFilters([]);
    setOnClearFilters(true);
    setOnClearData(true);
  };

  const handleClearChecks = () => {
    setOnClearFilters(false);
  };

  return (
    <>
      <Header classNameHeader='flex justify-center'>
        <SideDrawer classNameContainer='bg-black'>
          <CharactersControlPanel
            searchInput={searchInput}
            setSearchInput={setSearchInput}
            searchTitle={'Search by name'}
            searchPlaceholder={'Type a character name'}
            setOnClearData={setOnClearData}
            isEmptyData={!isFetching && filteredCharacters?.length === 0}
            emptyDataLiteral={EMPTY_DATA_LITERAL_LIST}
            orderTitle='Order results'
            onOrderChange={(event) => orderHandler(event)}
            orderOptions={Object.values(FetchingOrder)}
            orderLiterals={SELECT_ORDER_LITERALS}
            filtersTitle='Filter results:'
            filtersOptions={Object.values(FilterCriteria)}
            filtersLiterals={filterLiterals}
            setFilters={setFilters}
            setOnClearChecks={handleClearChecks}
            onClearChecks={onClearFilters}
          />
        </SideDrawer>
      </Header>
      <Container element={'main'}>
        <section
          className='relative z-0 bg-black py-10 mb-4 bg-hero-image bg-center bg-cover
        before:absolute  before:content[""] before:h-full before:w-full  before:bg-gradient-to-b before:from-trans-0.75-black before:from-50% before:left-0 before:top-0 before:z-[-1]'
        >
          <Container element={'div'}>
            <div className='w-full text-white text-center mb-4'>
              <h2 className='text-2xl font-semibold text-white text-center uppercase mb-1'>
                Marvel Characters
              </h2>
              <p className='text-sm text-center'>
                Get hooked on a hearty helping of heroes and villains from the humble House of
                Ideas!
              </p>

              {!hasListDefaults && (
                <>
                  <p className='text-sm mb-2'>
                    Results for
                    {searchInput && <strong> {searchInput}</strong>}
                    {filters && <strong> {filters}</strong>}
                    {order && <strong> {order}</strong>}
                  </p>
                  <button
                    className='text-sm border border-white p-1 active-border focus-visible-border'
                    onClick={handleClear}
                  >
                    CLEAR
                  </button>
                </>
              )}
            </div>
            <CharactersControlPanel
              isDesktop
              searchInput={searchInput}
              setSearchInput={setSearchInput}
              searchTitle={'Search by name'}
              searchPlaceholder={'Type a character name'}
              setOnClearData={setOnClearData}
              isEmptyData={!isFetching && filteredCharacters?.length === 0}
              emptyDataLiteral={EMPTY_DATA_LITERAL_LIST}
              orderTitle='Order results'
              onOrderChange={(event) => orderHandler(event)}
              orderOptions={Object.values(FetchingOrder)}
              orderLiterals={SELECT_ORDER_LITERALS}
              filtersTitle='Filter results:'
              filtersOptions={Object.values(FilterCriteria)}
              filtersLiterals={filterLiterals}
              setFilters={setFilters}
              setOnClearChecks={handleClearChecks}
              onClearChecks={onClearFilters}
            />
          </Container>
        </section>
        {isError && <h2>Oooops, there&apos;s an unexpected error...try reloading again!</h2>}

        {isFetching && !isFetchingNextPage && (
          <RingLoader
            color={MARVEL_RED}
            size={LOADER_SIZE}
            className='mx-auto my-6'
            role='alert'
            aria-label='Characters List is loading'
            aria-busy='true'
            aria-live='polite'
          />
        )}

        {!isFetching && (
          <Container>
            {filteredCharacters?.length === 0 ? (
              <h3 className='text-center'>{EMPTY_DATA_LITERAL_LIST}</h3>
            ) : (
              <CharactersList characters={filteredCharacters} />
            )}
          </Container>
        )}

        {hasNextPage && (
          <div ref={ref}>
            <RingLoader
              color={MARVEL_RED}
              size={LOADER_SIZE}
              className='mx-auto my-6'
              role='alert'
              aria-label='Characters List is loading'
              aria-busy='true'
              aria-live='polite'
            />
          </div>
        )}
      </Container>
      <Footer />
    </>
  );
};

export default Characters;
