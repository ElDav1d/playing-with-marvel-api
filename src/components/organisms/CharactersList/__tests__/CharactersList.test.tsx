import { render, screen } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import CharactersList from '../CharactersList';
import { mockCharactersAZ, mockCharactersZA } from '@/components/pages/Characters/mockCharacters';

describe(CharactersList, () => {
  it('renders a list of characters', () => {
    // ARRANGE
    const filters = [''];

    // ACT
    render(
      <Router>
        <CharactersList filters={filters} characters={mockCharactersAZ} />
      </Router>,
    );

    const headings = screen.getAllByRole('heading');
    expect(headings.length).toBe(4);
  });

  it('renders a filtered list of characters with description', () => {
    // ARRANGE
    const filters = ['withDescription'];

    // ACT
    render(
      <Router>
        <CharactersList filters={filters} characters={mockCharactersAZ} />
      </Router>,
    );

    // ASSERT
    const headings = screen.getAllByRole('heading');
    expect(headings.length).toBe(2);
  });

  it('renders filtered list of characters with image', () => {
    // ARRANGE
    const filters = ['withImage'];

    // ACT
    render(
      <Router>
        <CharactersList filters={filters} characters={mockCharactersAZ} />
      </Router>,
    );

    // ASSERT
    const headings = screen.getAllByRole('heading');
    expect(headings.length).toBe(2);
  });

  it('renders a filtered list of characters with description and image', () => {
    // ARRANGE
    const filters = ['withDescription', 'withImage'];

    // ACT
    render(
      <Router>
        <CharactersList filters={filters} characters={mockCharactersAZ} />
      </Router>,
    );

    // ASSERT
    const headings = screen.getAllByRole('heading');
    expect(headings.length).toBe(1);
  });

  it('renders the full list of characters after removing the filters', () => {
    // ARRANGE
    const initFilters = ['withDescription', 'withImage'];

    const { rerender } = render(
      <Router>
        <CharactersList filters={initFilters} characters={mockCharactersAZ} />
      </Router>,
    );

    // ACT
    const noFilters = [''];
    rerender(
      <Router>
        <CharactersList filters={noFilters} characters={mockCharactersAZ} />
      </Router>,
    );

    // ASSERT
    const headings = screen.getAllByRole('heading');
    expect(headings.length).toBe(4);
  });

  it('renders a filtered list of characters with description after reordering', () => {
    // ARRANGE
    const initFilters = ['withDescription'];

    const { rerender } = render(
      <Router>
        <CharactersList filters={initFilters} characters={mockCharactersAZ} />
      </Router>,
    );

    // ACT
    rerender(
      <Router>
        <CharactersList filters={initFilters} characters={mockCharactersZA} />
      </Router>,
    );

    // ASSERT
    const headings = screen.getAllByRole('heading');
    expect(headings.length).toBe(2);
  });

  it('renders a filtered list of characters with description after adding more', () => {
    // ARRANGE
    const initFilters = ['withDescription'];
    const initStack = mockCharactersAZ;

    const { rerender } = render(
      <Router>
        <CharactersList filters={initFilters} characters={initStack} />
      </Router>,
    );

    // ACT
    const biggerStack = [...initStack, ...mockCharactersZA];

    rerender(
      <Router>
        <CharactersList filters={initFilters} characters={biggerStack} />
      </Router>,
    );

    // ASSERT
    const headings = screen.getAllByRole('heading');
    expect(headings.length).toBe(4);
  });
});
