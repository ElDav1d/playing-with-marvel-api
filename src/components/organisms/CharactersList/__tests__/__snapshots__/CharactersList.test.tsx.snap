// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the list of characters and matches  snapshot 1`] = `
<DocumentFragment>
  <ul
    aria-label="characters list"
    aria-live="polite"
    class="grid gap-3 grid-flow-row grid-cols-auto-min-max-120-auto md:grid-cols-auto-min-max-185-auto"
  >
    <li
      aria-label="Animal"
      class="bg-black text-white overflow-hidden relative z-0
      after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
      
    "
    >
      <a
        aria-labelledby="1011176"
        class="block group"
        href="/character/1011176/animal"
      >
        <picture
          class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
        />
        <div
          class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-s before:duration-l before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
        >
          <h3
            class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-2"
          >
            Animal
          </h3>
          <p
            class="text-xs leading-5 line-clamp-4"
          >
            Xorn, whose mutation manifested as a tiny, powerful star inside his head, was brainwashed into believing he was the reincarnation of Magneto.
          </p>
        </div>
      </a>
    </li>
    <li
      aria-label="Braineater"
      class="bg-black text-white overflow-hidden relative z-0
      after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
      
    "
    >
      <a
        aria-labelledby="1011177"
        class="block group"
        href="/character/1011177/braineater"
      >
        <picture
          class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
        />
        <div
          class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-s before:duration-l before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
        >
          <h3
            class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-2"
          >
            Braineater
          </h3>
          <p
            class="text-xs leading-5 line-clamp-4"
          >
            Xorn, whose mutation manifested as a tiny, powerful star inside his head, was brainwashed into believing he was the reincarnation of Magneto.
          </p>
        </div>
      </a>
    </li>
    <li
      aria-label="Crusher"
      class="bg-black text-white overflow-hidden relative z-0
      after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
      
    "
    >
      <a
        aria-labelledby="1011178"
        class="block group"
        href="/character/1011178/crusher"
      >
        <picture
          class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
        />
        <div
          class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-s before:duration-l before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
        >
          <h3
            class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-2"
          >
            Crusher
          </h3>
        </div>
      </a>
    </li>
    <li
      aria-label="Destructor"
      class="bg-black text-white overflow-hidden relative z-0
      after:absolute after:content[\\"\\"] after:z-1 after:bottom-0 after:right-0 after:border-8 after:border-t-transparent after:border-r-white after:border-b-white after:border-l-transparent
      
    "
    >
      <a
        aria-labelledby="1011179"
        class="block group"
        href="/character/1011179/destructor"
      >
        <picture
          class="relative after:absolute after:content[\\"\\"] after:bg-red after:h-1 after:w-full after:left-0 after:-bottom-0.5"
        />
        <div
          class="relative z-0 h-40 p-3 before:absolute before:transition-[max-height] before:ease-in-out before:delay-s before:duration-l before:content[\\"\\"] before:h-[105%] before:max-h-[0px] before:bg-red before:w-full before:left-0 before:-top-0.5 before:z-[-1] group-hover:before:max-h-[300px]"
        >
          <h3
            class="mb-2 font-semibold uppercase text-sm leading-5 line-clamp-2"
          >
            Destructor
          </h3>
        </div>
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
